### DISCUSSION TITLE
Go 语言中的 make 和 new · /golang-make-and-new

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#562] [Announcements]
https://draveness.me/golang-make-and-new

**Comment by abelsu7** [comment]
感谢博主持续输出，受教了

**Comment by zhijiay** [comment]
请教博主的画图工具是什么

**Comment by Hale-ma** [comment]
> @ZhijiaYang
> 请教博主的画图工具是什么
> 

看起来像是 xmind 吧

**Comment by ghost** [comment]
博主好，请问博主用的这个 jekyll 主题，您的网站源代码在哪存放呢，想参考一下相关设置

**Comment by draveness** [comment]
> 博主好，请问博主用的这个 jekyll 主题，您的网站源代码在哪存放呢，想参考一下相关设置

源代码私有的

**Comment by g10guang** [comment]
所以 new 就是创建在堆上的内存？还是说逃逸分析，如果不会逃逸就创建在栈上？

**Comment by draveness** [comment]
> @g10guang
> 所以 new 就是创建在堆上的内存？还是说逃逸分析，如果不会逃逸就创建在栈上？

`new` 也会有逃逸分析

**Comment by max-min** [comment]
new和var 有啥区别呢？ 
new作用域外没有使用，可能在栈上分配。
var如果在外面引用也是被分配在堆上。 
new的作用是啥？

**Comment by draveness** [comment]
> new和var 有啥区别呢？ 
> new作用域外没有使用，可能在栈上分配。
> var如果在外面引用也是被分配在堆上。 
> new的作用是啥？

没啥大区别，new 返回的是个指针

**Comment by nochokolate** [comment]
> new 的作用是根据传入的类型在堆上分配一片内存空间并返回指向这片内存空间的指针；

博主，这个地方是不是忘记把“在堆上”这三个子给去掉了？

---

*2020-05-20 UPDATES：已修复*

**Comment by waittttting** [comment]
> cmd/compile/internal/gc.callnew 函数会将关键字转换成 ONEWOBJ 类型的节点

> 无论是直接使用 new，还是使用 var 初始化变量，它们在编译器看来就是 ONEW 和 ODCL 节点

请教一下 博主 上述 是 编译器 对 `new` 这个关键字的 不同的处理方式吗，`new` 是被转换成了 `ONEWOBJ` 还是 `ONEW`

---

*2020-05-24 UPDATES: 已修复*

**Comment by draveness** [comment]
> > cmd/compile/internal/gc.callnew 函数会将关键字转换成 ONEWOBJ 类型的节点
> 
> > 无论是直接使用 new，还是使用 var 初始化变量，它们在编译器看来就是 ONEW 和 ODCL 节点
> 
> 请教一下 博主 上述 是 编译器 对 `new` 这个关键字的 不同的处理方式吗，`new` 是被转换成了 `ONEWOBJ` 还是 `ONEW`

`ONEWOBJ` 后面那个应该写错了

**Comment by MoeexT** [comment]
官方文档中make返回的是原始类型而不是其指针啊

**Comment by draveness** [comment]
> 官方文档中make返回的是原始类型而不是其指针啊

给一下 Ref 和文中有问题的片段？

**Comment by JustForWL** [comment]
> @draveness
> > 官方文档中make返回的是原始类型而不是其指针啊
> 
> 给一下 Ref 和文中有问题的片段？

hash 是一个指向 runtime.hmap 结构体的指针；
ch 是一个指向 runtime.hchan 结构体的指针；
hash和ch都不是指针

**Comment by KylinMountain** [comment]
我感觉这几节关键字看下来没有啥收获，反而看了有一种困惑的感觉。感觉只需要像这一节简单说明一下，存在的坑，如何避免就好了。你像defer和panic那节，有哪个会在defer中写panic，defer中套defer，感觉这么写，要被开除了。

**Comment by draveness** [comment]
> 我感觉这几节关键字看下来没有啥收获，反而看了有一种困惑的感觉。感觉只需要像这一节简单说明一下，存在的坑，如何避免就好了。你像defer和panic那节，有哪个会在defer中写panic，defer中套defer，感觉这么写，要被开除了。

确实不会这么写，不过需要注意的是这不是一个实践类的系列文章，这里讨论的是语言底层的实现原理

**Comment by Northward712** [comment]
new分配在heap上还是在stack上首要看的应该是他是否超过内存管理中的32KB的限制的吧。然后才是看是否是逃逸。也就是说new不是肯定分配在heap上的

**Comment by draveness** [comment]
> new分配在heap上还是在stack上首要看的应该是他是否超过内存管理中的32KB的限制的吧。然后才是看是否是逃逸。也就是说new不是肯定分配在heap上的

这个结论是从哪里得到的呢，把下面的代码编译成 SSA 中间代码看一下？

```go
package main

import "fmt"

type Test struct {
	i int
}

func main() {
	a := new(Test)
	fmt.Println(a)
}

$ GOSSAFUNC=main go build main.go
```

![image](https://user-images.githubusercontent.com/6493255/93965990-1ce65400-fd96-11ea-9568-03b816930cc2.png)

**Comment by Northward712** [comment]
> @draveness
> > new分配在heap上还是在stack上首要看的应该是他是否超过内存管理中的32KB的限制的吧。然后才是看是否是逃逸。也就是说new不是肯定分配在heap上的
> 
> 这个结论是从哪里得到的呢，把下面的代码编译成 SSA 中间代码看一下？
> 
> ```go
> package main
> 
> import "fmt"
> 
> type Test struct {
> 	i int
> }
> 
> func main() {
> 	a := new(Test)
> 	fmt.Println(a)
> }
> 
> $ GOSSAFUNC=main go build main.go
> ```
> 
> ...
> 

https://mp.weixin.qq.com/s/Slf7J6ZP4Uz9ukRc-JORyg  这里看的

**Comment by draveness** [comment]
> > @draveness
> > > new分配在heap上还是在stack上首要看的应该是他是否超过内存管理中的32KB的限制的吧。然后才是看是否是逃逸。也就是说new不是肯定分配在heap上的
> > 
> > 
> > 这个结论是从哪里得到的呢，把下面的代码编译成 SSA 中间代码看一下？
> > ```go
> > package main
> > 
> > import "fmt"
> > 
> > type Test struct {
> > 	i int
> > }
> > 
> > func main() {
> > 	a := new(Test)
> > 	fmt.Println(a)
> > }
> > 
> > $ GOSSAFUNC=main go build main.go
> > ```
> > 
> > 
> 
> https://mp.weixin.qq.com/s/Slf7J6ZP4Uz9ukRc-JORyg 这里看的

这里也没讲 new 关键字的原理，逃逸分析和 new 关键字的编译是两个过程，后面讲到了逃逸分析，具体可以看下栈内存管理一节

**Comment by Northward712** [comment]
> > > @draveness
> > > > new分配在heap上还是在stack上首要看的应该是他是否超过内存管理中的32KB的限制的吧。然后才是看是否是逃逸。也就是说new不是肯定分配在heap上的
> > > 
> > > 
> > > 这个结论是从哪里得到的呢，把下面的代码编译成 SSA 中间代码看一下？
> > > ```go
> > > package main
> > > 
> > > import "fmt"
> > > 
> > > type Test struct {
> > > 	i int
> > > }
> > > 
> > > func main() {
> > > 	a := new(Test)
> > > 	fmt.Println(a)
> > > }
> > > 
> > > $ GOSSAFUNC=main go build main.go
> > > ```
> > 
> > 
> > https://mp.weixin.qq.com/s/Slf7J6ZP4Uz9ukRc-JORyg 这里看的
> 
> 这里也没讲 new 关键字的原理，逃逸分析和 new 关键字的编译是两个过程，后面讲到了逃逸分析，具体可以看下栈内存管理一节







您好 不去说逃逸分析，只说堆分配和栈分配的问题
标准库这个函数上面注释说的是32KB限制了
而且代码内也对maxSmallSize做了判断
如果没理解错 是不是表示是32KB以下会在栈上分配内存
 不知道我理解的是否有问题
<img width="821" alt="截屏2020-12-30 15 10 56" src="https://user-images.githubusercontent.com/37608723/103335823-3a237080-4ab1-11eb-9179-d40f92507d7d.png">
<img width="566" alt="截屏2020-12-30 15 16 15" src="https://user-images.githubusercontent.com/37608723/103336076-f8df9080-4ab1-11eb-8447-5a066e621637.png">

**Comment by draveness** [comment]
> 您好 不去说逃逸分析，只说堆分配和栈分配的问题
> 标准库这个函数上面注释说的是32KB限制了
> 而且代码内也对maxSmallSize做了判断
> 如果没理解错 是不是表示是32KB以下会在栈上分配内存

这里说的是 32KB 大小以下的会用处理器本地的 freeList 分配内存，也是在栈上只是分配速度会快一些

**Comment by dillanzhou** [comment]
“如果变量不会逃逸到堆上，这些节点在这一阶段都会被 cmd/compile/internal/gc.walkstmt 转换成通过 runtime.newobject 函数并在堆上申请内存”。为啥不会逃逸到堆上还要在堆上申请内存？

**Comment by draveness** [comment]
> “如果变量不会逃逸到堆上，这些节点在这一阶段都会被 cmd/compile/internal/gc.walkstmt 转换成通过 runtime.newobject 函数并在堆上申请内存”。为啥不会逃逸到堆上还要在堆上申请内存？

没太懂你的问题

**Comment by dillanzhou** [comment]
> > “如果变量不会逃逸到堆上，这些节点在这一阶段都会被 cmd/compile/internal/gc.walkstmt 转换成通过 runtime.newobject 函数并在堆上申请内存”。为啥不会逃逸到堆上还要在堆上申请内存？
> 
> 没太懂你的问题

如果变量不会逃逸到堆上，那么变量在栈上分配就可以了，为什么要到堆上去申请？不应该是会逃逸到堆上的变量才需要到堆上申请内存？

**Comment by draveness** [comment]
> @dillanzhou
> > > “如果变量不会逃逸到堆上，这些节点在这一阶段都会被 cmd/compile/internal/gc.walkstmt 转换成通过 runtime.newobject 函数并在堆上申请内存”。为啥不会逃逸到堆上还要在堆上申请内存？
> > 
> > 没太懂你的问题
> 
> 如果变量不会逃逸到堆上，那么变量在栈上分配就可以了，为什么要到堆上去申请？不应该是会逃逸到堆上的变量才需要到堆上申请内存？

这里是 typo，已经修复了

**Comment by lefeck** [comment]
错误： 需要注意的是，无论是直接使用 new，还是使用 var 初始化变量，它们在编译器看来都是 ONEWOBJ 和 ODCL 节点

case ONEW：和 case ODCL:   


上面改成：是ONEW和 ODCL 节点，不是ONEWOBJ 吧？

---

*2021-05-05 UPDATES: 已修复*

**Comment by vchive** [comment]
make返回的不是指针，建议更改
```go
slice 是一个包含 data、cap 和 len 的结构体 reflect.SliceHeader；
hash 是一个指向 runtime.hmap 结构体的指针；
ch 是一个指向 runtime.hchan 结构体的指针；
```

**Comment by NayelyA** [comment]
```golang
package main

import "fmt"

func main() {
	slice := make([]int, 0, 100)
	hash := make(map[int]bool, 10)
	ch := make(chan int, 5)

	fmt.Println("slice->>>", slice)//slice->>> []
	fmt.Println("hash->>>", hash)//hash->>> map[]
	fmt.Println("ch->>>", ch)//ch->>> 0xc00012c000(chan string)
}
```
> hash 是一个指向 runtime.hmap 结构体的指针；
> ch 是一个指向 runtime.hchan 结构体的指针；
文中这里描述的不太准确

**Comment by MoCuishle28** [comment]
博主好，上面提到用 make 初始化 map 和 chan 返回的是指针。但是根据 https://pkg.go.dev/builtin#make 这里的说法，make 返回的似乎是和第一个参数相同的类型，而非指针？

**Comment by unbreakablehf** [comment]
请问博主 cmd/compile/internal/gc.callnew 这个超链接指向的代码在go版本1.17,1.18 哪个源码包里？
go1.16版本里面有， 1.17以上的包结构变化这么大？

再请教一个问题 点击这个超链接【cmd/compile/internal/gc.callnew】，是怎么跳转到这个链接的？  
https://github.com/golang/go/blob/41d8e61a6b9d8f9db912626eb2bbc535e929fefc/src/cmd/compile/internal/gc/walk.go#L2035
后台代码逻辑指定这个 【41d8e61a6b9d8f9db912626eb2bbc535e929fefc 】commmit？
